gdrive

```diff
@@ -91,22 +93,36 @@ def upload(service, filepath):
         created_file = service.files().create(body=file_metadata,
                                               media_body=media,
                                               fields='id')
-        response = None
+        upload_loop(created_file)
+    except Exception as err:
+        print('An error occurred while uploading the file.')
+        print(err)
+
+def upload_loop(created_file):
+    try:
+        last_size = 0.0
+        last_time = start_time = time.time()
+        response = logger = None
         print("Uploading 0%", end='\r')
         while response is None:
             status, response = created_file.next_chunk()
-            if status:
-                print("Uploading %d%% - %s/%s" % interpret_progress(status), end='\r')
-
-        print("Uploading 100%")
-        print('File ID: %s' % response.get('id'))
+            # Does not log if the previous logging is not done yet
+            if status and (not logger or not logger.isAlive()):
+                logger = threading.Thread(target=interpret_progress,
```

Comment:
you probably want `logger.join()` somewhere so the thread finishes gracefully

Reply:
With .join(), the main thread would wait for the logging to end, but, in my view, we can just skip a log if it's going that fast.

Reply:
As discussed, this has been remade into a Logger class with a ConflatedChannel.